File: wordracer-scala/src/main/java/com/codingnirvana/wordracer/Runner.java
package com.codingnirvana.wordracer;

import com.codingnirvana.wordracer.impl.DemoWordRacer;
import com.codingnirvana.wordracer.samples.*;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Runner {


    public static void main(String[] args) throws IOException {

        List<String> words = readWordsList();

        WordRacer dRacer = new JDemoWordRacer();

        if (args.length == 1) {
            String type = args[0].trim();
            if (type.equals("S")) {
                dRacer = new SlowWordRacer();
            } else if (type.equals("R")) {
                dRacer = new RogueWordRacer();
            } else {
                dRacer = new JDemoWordRacer();
            }

        }

        Scanner input = new Scanner(System.in);
        char firstLetter = input.nextLine().charAt(0);

        boolean myTurn = input.nextLine().charAt(0) == '1';

        dRacer.initGameBoard(firstLetter);
        char letter;
        for (int i = 0; i < 48; i++) {

            if (myTurn) {
                Result result = dRacer.pickLetter();
                System.out.println(result.letter + " " + result.position);
            } else {
                letter = input.nextLine().charAt(0);

                int position = dRacer.pickPosition(letter);
                System.out.println(position);
                System.out.flush();
            }

            myTurn = !myTurn;
        }
    }

    private static List<String> readWordsList() {
        List<String> words = new ArrayList<String>();
        Scanner scanner = new Scanner(Runner.class.getResourceAsStream("/words.dat"));
        while (scanner.hasNext()) {
            String line = scanner.nextLine();
            words.add(line);
        }
        return words;
    }

}

File: wordracer-scala/src/main/scala/com/codingnirvana/wordracer/Runner.scala
package com.codingnirvana.wordracer

import com.codingnirvana.wordracer.impl.DemoWordRacer

import scala.io.StdIn

object Runner {

  def main(args: Array[String]) = {
    val stream = getClass.getResourceAsStream("/words.dat")
    val words = scala.io.Source.fromInputStream(stream).getLines.toIndexedSeq.sorted

    val wRacer = new DemoWordRacer(words)

    wRacer.initGameBoard(StdIn.readChar())

    val myTurn = StdIn.readChar() == '1'

    (0 until 48).foldLeft(myTurn) { case (turn, i) =>
      if (turn) {
        val result = wRacer.pickLetter
        println(result.letter + " " + result.position)
        Console.flush()
      } else {
        val letter = StdIn.readChar()
        val position = wRacer.pickPosition(letter)
        println(position)
        Console.flush()
      }
      !turn
    }

    System.exit(0)
  }

}
